torch/profiler/strobelight_cli_function_profiler.py

```diff
@@ -0,0 +1,311 @@
+    link = pid_namespace_link(pid)
+    return int(link[link.find("[") + 1 : -1])
+
+
+def command_to_string(command):
+    s = ""
+    for item in command:
+        s += " "
+        s += item
+    return s
+
+
+class StrobelightCLIFunctionProfiler:
+    """
+    Note: this is a meta only tool.
+
+    StrobelightCLIFunctionProfiler can be used to profile a python function and
+    generate a strobelight link with the results. It works on meta servers but
+    does not requries an fbcode target.
+    When stop_at_error is false(default), error during profiling does not prevent
+    the work function from running.
+
+    Check function_profiler_example.py for an example.
+    """
+
+    # This lock is used to make sure only one thread is running the profiler at any point.
+    lock = Lock()
+
+    def __init__(
+        self,
+        *,
+        stop_at_error=False,
+        max_profile_duration_sec=60 * 10,
+        sample_each=1e7,  # sample each sample_each cycles.
+        run_user_name="pytorch-strobelight-ondemand",
+        timeout_wait_for_running_sec=60,
+        timeout_wait_for_finished_sec=60,
+        recorded_env_variables=None,
+        sample_tags=None,
+        stack_max_len=127,
+        async_stack_max_len=127,
+    ):
+        self.stop_at_error = stop_at_error
+        self.max_profile_duration_sec = max_profile_duration_sec
+        self.sample_each = sample_each
+        self.run_user_name = run_user_name
+        self.timeout_wait_for_running_sec = timeout_wait_for_running_sec
+        self.timeout_wait_for_finished_sec = timeout_wait_for_finished_sec
+        # Results of the most recent run.
+        # Tracks the strobelight run id of the most recent run
+        self.current_run_id = None
+        self.sample_tags = sample_tags
+
+    def _run_async(self):
+        processId = os.getpid()
+        namespace = pid_namespace(processId)
+        command = [
+            "strobeclient",
+            "run",
+            "--profiler",
+            "pyperf",
+            "--event",
+            "cycles",
+            "--async",
+            "--sample-interval",
+            str(int(self.sample_each)),
+            "--duration-ms",
+            str(int(self.max_profile_duration_sec * 1000)),
+            "--pid",
+            f"{namespace}:{processId}",
+        ]
+
+        if self.sample_tags:
+            command.append("--sample-tags")
+            add_comma = False
+            for item in self.sample_tags:
+                if add_comma:
+                    command.append(",")
+                command.append(item)
+                add_comma = True
+
+        logger.debug(f"running command:{command_to_string(command)}")
+
+        result = subprocess.run(command, capture_output=True)
+
+        if result.returncode != 0:
+            raise StrobelightCLIProfilerError(
+                "failed to start strobelight profiling, error in run_async"
+            )
+
+        for line in result.stderr.split(b"\n"):
```

Comment:
nit:
```
for line in result.stderr.splitlines():
```

but less nit:
Replace this whole `for` loop with a single `re.search` for the pattern you're looking for - which should include the match group - something like this:
```
if match := re.search(rb"INFO Run Id:.*(-?\d+)$", line):
  self.current_run_id = int(match.group(1))
```

Reply:
mhmm let me look i think the reason i added the loop is that logs get pritned all in one line (stream of bytes) but I can decode result first to avoid the loop.